### PR TITLE
Move ClipExtension to @deck.gl/extensions

### LINKED ISSUE #5725: Clip layers
#### Target Use case
Clip layers by bounds defined in JS, instead of requiring to clip the original data.

Usecases:
- Original MVTLayer / GeoJsonLayer usecase
- BitmapLayer with experimental _imageCoordinateSystem = COORDINATE_SYSTEM.LNGLAT
- potentially any other layer, if the original data can't be clipped easily at the source

#### Proposed feature
There is private [ClipExtension](https://github.com/visgl/deck.gl/blob/master/modules/geo-layers/src/mvt-layer/clip-extension.js) already. As discussed on Slack, the proposal is to move it to @deck.gl/extensions, so that it can be used by any layer.

Without ClipExtension:

Screen Shot 2021-05-05 at 8 48 53

With ClipExtension and clipBounds = [-180, -85.06, 180, 85.06]:

Screen Shot 2021-05-05 at 8 49 11

#### To Do List 
  - [ ] Add label and assign to milestone 
  - [x] Coding
  - [ ] Doc update
  - [ ] What’s new update 
  - [ ] Test

### PR DESCRIPTION
**Pull request by zakjan**
Fixes #5725

### CODE REVIEW

**Comment by Pessimistress** [comment]
Thanks!

TODO:

- [ ] add ClipExtension documentation
- [ ] update MVTLayer documentation to mention the new dependency
- [ ] add to upgrade guide that geo-layers now has new dependency on extensions

**Comment by ibgreen** [comment]
@zakjan This looks good - are you able to address the remaining items @Pessimistress listed or do you need some pointers?

**Comment by zakjan** [comment]
Docs added for ClipExtension and clipBounds, added as dependency for MVTLayer, added to upgrade guide. Let me know if it fits with common docs writing style.

I would need help with documenting clipByInstance though. It's not used anywhere, I'm not sure what's the usecase.

Also, I found that similar implicit clipping is used in HeatmapLayer. Is it duplicate, or a different usecase? https://github.com/visgl/deck.gl/blob/26bcb63f238158ff2bdbf629787092bd92d6c88d/modules/aggregation-layers/src/heatmap-layer/heatmap-layer.js#L401-L407

**Comment by ibgreen** [comment]
> Also, I found that similar implicit clipping is used in HeatmapLayer. Is it duplicate, or a different usecase?

Interesting observation. I am not current on the heatmap layer code, but generally I have a sense that it has been a rather big separate piece of code that is heavier to maintain than I would prefer, and it would merit from being more "componentized" and aligned with other parts of the code. Using the ClipExtension there could be a good inroad, if it makes sense.

I can imagine that some of the weather focused layer work you are setting out to do would have parallels with the heatmap layer (rendering into textures etc) and there could be multiple refactor / code sharing opportunities for someone who was able to get familiar with the heatmap layer code base.

Definitely a separate PR. @Pessimistress likely has more insights as to what would be feasible.

**Comment by Pessimistress** [comment]
HeatmapLayer does not perform clipping in the shader. The texture is dynamically generated to fill the viewport. The referenced code is totally unrelated to the use case of the ClipExtension.

File: docs/api-reference/extensions/clip-extension.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
+
+# ClipExtension
+
+The `ClipExtension` adds support for clipping rendered layers by rectangular bounds.
+
+```js
+import {BitmapLayer} from '@deck.gl/layers';
+import {ClipExtension} from '@deck.gl/extensions';
+
+const layer = new BitmapLayer({
```
Comment (Pessimistress) [inline docs/api-reference/extensions/clip-extension.md:10]: I recommend using a GeoJsonLayer as example. The code here does not illustrate why clipping is useful.

File: docs/api-reference/extensions/clip-extension.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+```
+
+## Installation
+
+To install the dependencies from NPM:
+
+```bash
+npm install deck.gl
+# or
+npm install @deck.gl/core @deck.gl/layers @deck.gl/extensions
```
Comment (Pessimistress) [inline docs/api-reference/extensions/clip-extension.md:24]: layers is not a dependency of extensions

File: docs/api-reference/extensions/clip-extension.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+new ClipExtension({});
+```
+
+To use pre-bundled scripts:
+
+```html
+<script src="https://unpkg.com/deck.gl@^7.0.0/dist.min.js"></script>
+<!-- or -->
+<script src="https://unpkg.com/@deck.gl/core@^7.0.0/dist.min.js"></script>
+<script src="https://unpkg.com/@deck.gl/layers@^7.0.0/dist.min.js"></script>
```
Comment (Pessimistress) [inline docs/api-reference/extensions/clip-extension.md:38]: Ditto

File: docs/api-reference/extensions/clip-extension.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+## Constructor
+
+```js
+new ClipExtension();
+```
+
+## Layer Properties
+
+When added to a layer via the `extensions` prop, the `ClipExtension` requires the `clipBounds` prop of the layer to contain the rectangular bounds to be used for clipping the rendered region.
```
Comment (Pessimistress) [inline docs/api-reference/extensions/clip-extension.md:54]: Each prop should have its own header, follow the convention in https://github.com/visgl/deck.gl/blob/master/docs/api-reference/extensions/data-filter-extension.md#layer-properties

**Review by Pessimistress — COMMENTED** [review]
`clipByInstance` controls whether an object is clipped by its anchor (e.g. icon, point) or by its geometry (e.g. path, polygon). If not specified, it is deduced from whether there is an attribute called `instancePositions`. The user might want to override this behavior if:
- the anchor attribute goes by some other name
- to clip an anchored layer by geometry, like the text layer

**Comment by zakjan** [comment]
Thanks @Pessimistress, initially I used Fp64Extension as a template. Docs updated.

**Review by Pessimistress — APPROVED** [review]

**Pull request merged**